contracts/test/veLON.t.sol

```diff
@@ -0,0 +1,265 @@
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, MAX_LOCK_TIME);
+        vm.prank(user);
+        vm.assume(_amount <= 50 * 1e18);
+        vm.assume(_amount > 0);
+        veLon.depositFor(tokenId, _amount);
+    }
+
+    function testExtendLock() public {
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, 1 weeks);
+        uint256 lockEndBefore = veLon.unlockTime(tokenId);
+
+        vm.prank(user);
+        veLon.extendLock(tokenId, 2 weeks);
+        uint256 lockEndAfter = veLon.unlockTime(tokenId);
+
+        require((lockEndAfter - lockEndBefore) == 1 weeks, "wrong time extended");
+    }
+
+    function testCannoExtendLockExpired() public {
+        uint256 lockTime = 1 weeks;
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, lockTime);
+
+        //pretend 1 week has passed and the lock expired
+        vm.warp(block.timestamp + 1 weeks);
+        vm.prank(user);
+        vm.expectRevert("Lock expired");
+        veLon.extendLock(tokenId, lockTime);
+    }
+
+    function testMerge() public {
+        uint256 _fromTokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, MAX_LOCK_TIME);
+
+        uint256 _toTokenId = _stakeAndValidate(other, DEFAULT_STAKE_AMOUNT, MAX_LOCK_TIME);
+        vm.prank(other);
+        veLon.approve(address(user), _toTokenId);
+
+        vm.prank(user);
+        veLon.merge(_fromTokenId, _toTokenId);
+
+        //check whether fromToken has burned
+        vm.expectRevert("ERC721: owner query for nonexistent token");
+        veLon.ownerOf(_fromTokenId);
+    }
+
+    /*********************************
+     *         Test: Withdraw        *
+     *********************************/
+
+    function testWithdraw() public {
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, 1 weeks);
+
+        stakerLon = BalanceSnapshot.take(user, address(lon));
+        lockedLon = BalanceSnapshot.take(address(veLon), address(lon));
+
+        //pretend 1 week has passed and the lock expired
+        vm.prank(user);
+        vm.warp(block.timestamp + 1 weeks);
+        veLon.withdraw(tokenId);
+        stakerLon.assertChange(int256(DEFAULT_STAKE_AMOUNT));
+        lockedLon.assertChange(int256(-(DEFAULT_STAKE_AMOUNT)));
+
+        //check whether token has burned after withdraw
+        vm.expectRevert("ERC721: owner query for nonexistent token");
+        veLon.ownerOf(tokenId);
+    }
+
+    function testWithdrawEarly() public {
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, 2 weeks);
+
+        stakerLon = BalanceSnapshot.take(user, address(lon));
+        lockedLon = BalanceSnapshot.take(address(veLon), address(lon));
+
+        //pretend 1 week has passed and the lock not expired
+        vm.warp(block.timestamp + 1 weeks);
+        vm.prank(user);
+        veLon.withdrawEarly(tokenId);
+        uint256 balanceChange = DEFAULT_STAKE_AMOUNT.mul(earlyWithdrawPenaltyRate).div(PENALTY_RATE_PRECISION);
```

Comment:
The `earlyWithdrawPenaltyRate` is mutable and we may change it to another number someday, which would make the test fail. Thus, it is better to get the value from the contract directly.